### PR TITLE
Add solution of day13

### PR DESCRIPTION
**Pull request by adpi2**
https://adventofcode.com/2021/day/13

### CODE REVIEW

File: src/day13.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day13"))(_.mkString)
+
+def part1(input: String): Int =
+  val (dots, folds) = parseInstructions(input)
+  dots.map(folds(0).apply).size
+
+def part2(input: String): String =
+  val (dots, folds) = parseInstructions(input)
+  val foldedDots = folds.foldLeft(dots)((dots, fold) => dots.map(fold.apply))
```
Comment (vincenzobaz) [inline src/day13.scala:25]: `dots` is shadowed in the `foldLeft`. You might consider another variable name

File: src/day13.scala [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  val (length, height) = (foldedDots.map(_.x).max + 1, foldedDots.map(_.y).max + 1)
+  val paper = Array.fill(height, length)('.')
+  for dot <- foldedDots do paper(dot.y)(dot.x) = '#'
+  
+  paper.map(_.mkString).mkString("\n")
+
+def parseInstructions(input: String): (Set[Dot], List[Fold]) =
+  val sections = input.split("\n\n")
+  val dots = sections(0).split('\n').toSet.map(Dot.parse)
+  val folds = sections(1).split('\n').toList.map(Fold.parse)
```
Comment (vincenzobaz) [inline src/day13.scala:36]: `linesIterator.map(...).toSet` might be more efficient

File: src/day13.scala [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  (dots, folds)
+
+def fold(along: Int)(point: Int): Int =
```
Comment (vincenzobaz) [inline src/day13.scala:39]: `fold` could be `private` to `enum Fold`.

File: src/day13.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+@main def part2(): Unit =
+  val answer = part2(readInput())
+  println(s"The answer is:\n$answer")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day13"))(_.mkString)
+
+def part1(input: String): Int =
+  val (dots, folds) = parseInstructions(input)
+  dots.map(folds(0).apply).size
```
Comment (vincenzobaz) [inline src/day13.scala:21]: Maybe `head` is more idiomatic on lists than `(0)`

**Review by vincenzobaz — APPROVED** [review]
It looks great! Just a few remarks

File: src/day13.scala (hunk 2 above)
Comment (adpi2) [reply src/day13.scala:36]: :+1:

File: src/day13.scala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+def parseInstructions(input: String): (Set[Dot], List[Fold]) =
+  val sections = input.split("\n\n")
+  val dots = sections(0).split('\n').toSet.map(Dot.parse)
+  val folds = sections(1).split('\n').toList.map(Fold.parse)
+  (dots, folds)
+
+def fold(along: Int)(point: Int): Int =
+  if point < along then point
+  else 2 * along - point
```
Comment (sjrd) [inline src/day13.scala:41]: My intuition here would be to write `along - (point - along)`. Of course these are equivalent, but I can only convince myself that `2 * along - point` is correct by verifying that it is equivalent to `along - (point - along)`. Perhaps add a comment or something?

File: src/day13.scala (hunk 1 above)
Comment (adpi2) [reply src/day13.scala:25]: Unfortunately I cannot find a better name.

Comment (julienrf) [reply src/day13.scala:25]: `foldedDotsAcc`? Not sure it is better :laughing:

File: src/day13.scala [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+def part2(input: String): String =
+  val (dots, folds) = parseInstructions(input)
+  val foldedDots = folds.foldLeft(dots)((dots, fold) => dots.map(fold.apply))
+  
+  val (width, height) = (foldedDots.map(_.x).max + 1, foldedDots.map(_.y).max + 1)
+  val paper = Array.fill(height, width)('.')
+  for dot <- foldedDots do paper(dot.y)(dot.x) = '#'
+  
+  paper.map(_.mkString).mkString("\n")
```
Comment (julienrf) [inline src/day13.scala:31]: Seems beautiful, I want to see the result!

File: src/day13.scala [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  val dots = sections(0).linesIterator.map(Dot.parse).toSet
+  val folds = sections(1).linesIterator.map(Fold.parse).toList
+  (dots, folds)
+
+case class Dot(x: Int, y: Int)
+
+object Dot:
+  def parse(line: String): Dot =
+    line match
+      case s"$x,$y" => Dot(x.toInt, y.toInt)
```
Comment (julienrf) [inline src/day13.scala:44]: :star_struck:

**Pull request merged**